File: Data Structures & Algorithms/maximum-product-subarray/submission-3.java
class Solution {
    public int maxProduct(int[] nums) {
        int maxProd = nums[0];
        int currMin = 1;
        int currMax = 1;

        for(int n : nums)
        {
            int temp = n * currMin;
            currMin = Math.min(n, Math.min(n * currMin, n * currMax));
            currMax = Math.max(n, Math.max(temp, n * currMax));
            maxProd = Math.max(currMax, maxProd);

        }

        return maxProd;
        
    }
}

File: Data Structures & Algorithms/maximum-subarray/submission-3.py
class Solution:
    def maxSubArray(self, nums: List[int]) -> int:
        if len(nums) < 1:
            return 0

        prefixSum = nums[0]
        largestSum = nums[0]

        for i in range(1, len(nums)):
            if prefixSum < 0:
                prefixSum = 0
            prefixSum = prefixSum + nums[i]
            largestSum = max(largestSum, prefixSum)

        return largestSum

        